feat(index.jsx): Updating to support react-cosmos3

scottdickerson opened the pull request:
test(proxy.spec & default.spec) Adding testcases to reach full code
coverage through unit tests (and pass build)
feat(package.json): updating peer deps and dev deps to react-cosmos3,
added enzyme to support new unit test

Commit:
feat(index.jsx): Updating to support react-cosmos3
test(proxy.spec): updated proxy.spec to match cosmos3
feat(package.json): updating peer deps and dev deps to react-cosmos3,
feat(cosmos.config.js & fixtures*) updated cosmos playground to support v3

NiGhTTraX commented:
@scottdickerson I think the unit test that checks the default values is unnecessary as we should be getting coverage from the acceptance tests. It's also better to not have the unit tests check more than 1 component at a time - in this case we're checking both the proxy and the `Docs` component.

Also, there's an [error](https://travis-ci.org/NiGhTTraX/react-docs-cosmos-proxy/builds/353953221#L3885) on Travis about compiling the Cosmos config. Not sure if anything's changed with Cosmos@3 related to this...

P.S.: Please don't commit the package lock for now, I think it should be added when the project is in a releasable state.

scottdickerson commented:
@NiGhTTraX I made all the changes you requested, but am now getting null in my nyc coverage json files and nyc is throwing an error trying to rollup the coverage reports.  Any ideas?
`emoving acceptance_playground_1         ... done
Removing network acceptance_default
/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/node_modules/yargs/yargs.js:1100
      else throw err
           ^

TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at SourceMaps.reloadCachedSourceMaps (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/lib/source-maps.js:37:10)
    at /Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/index.js:528:22
    at Array.map (<anonymous>)
    at NYC.loadReports (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/index.js:517:16)
    at NYC._getCoverageMapFromAllCoverageFiles (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/index.js:424:8)
    at NYC.report (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/index.js:442:18)
    at Object.exports.handler (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/lib/commands/report.js:38:7)
    at Object.runCommand (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/node_modules/yargs/lib/command.js:228:22)
    at Object.parseArgs [as _parseArgs] (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/node_modules/yargs/yargs.js:1013:30)
npm ERR! Test failed.  See above for more details.`

Commit:
test(acceptance): fixup the acceptance tests to support cosmos-3
fixup the acceptance tests to support cosmos-3
test(helpers): wait for the iframe to become visible,
test(setup): increase timeout waiting on browser to become visible
test(screenshots): update all screenshots

Commit:
code coverage is not collecting during acceptance tests,
improve unit test coverage to pass builds 100% coverage threshold
acceptance test automation handle slow travis build system wait for elements before attempting to click

scottdickerson commented:
@NiGhTTraX This is ready to go but FYI our acceptance tests are not generating code coverage reports (I added a warning to catch and report this case).  I went ahead and added the necessary unit tests to pass the code coverage threshold.

scottdickerson commented:
@NiGhTTraX ok it's ready for you to review...

NiGhTTraX commented on package.json:
```diff
@@ -28,8 +28,8 @@
     "test:acceptance:debug": "./tests/acceptance/debug.sh",
-    "_test:acceptance": "mocha --opts tests/acceptance/mocha.opts tests/acceptance/setup.js --recursive tests/acceptance/specs",
-    "_test:acceptance:debug": "mocha --opts tests/acceptance/mocha.debug.opts tests/acceptance/setup.js --recursive tests/acceptance/specs",
+    "_test:acceptance": "mocha --require ignore-styles --opts tests/acceptance/mocha.opts tests/acceptance/setup.js --recursive tests/acceptance/specs",
```
Please move the `--require` param to the opts file - this way we don't invalidate the Docker cache.

NiGhTTraX commented on playground/components/acceptance/.babelrc:
```diff
@@ -2,6 +2,7 @@
   "plugins": [
-    "react-docgen"
+    "react-docgen",
+    "istanbul"
```
The "acceptance" components for the playground don't need to be instrumented, they're basically fixtures. If the coverage reports are failing because of this that's fine, we'll fix it later.

NiGhTTraX commented on playground/cosmos.config.js:
```diff
@@ -1,16 +1,8 @@
-  ],
-
+module.exports = {
+  fileMatch: '**/{__fixture?(s)__,fixture?(s)}/**/*.{js,jsx,ts,tsx}',
```
Please remove the TypeScript extensions :)

NiGhTTraX commented on src/index.jsx:
```diff
@@ -60,3 +60,5 @@ export default ({
   }
 };
+
+export default createDocsProxy;
```
There's no need for this to be separated from the assignment since we assign static members using class fields. Please move it back up.

scottdickerson replied:
I was getting an eslint error I think on this because the function wasn't named.

NiGhTTraX commented on tests/acceptance/helpers.js:
```diff
@@ -12,6 +12,7 @@ export async function loadFixture(component, fixture) {
 
   // Switch focus to the only frame on the page, the one displaying the
   // component.
+  await browser.waitForVisible('iframe', 15 * 1000);
```
Why is this necessary? The tests setup already [waits](https://github.com/NiGhTTraX/react-docs-cosmos-proxy/blob/16e64d54dbb4c436233623363605156b0f777987/tests/acceptance/setup.js#L54) for the playground to load. Same question for the `checkForVisualChanges` change below.

scottdickerson replied:
I think the iframe was loading a little slower than the main playground header.  (this is my guess anyway, these timing issues were hard to resolve)

NiGhTTraX commented on tests/acceptance/setup.js:
```diff
@@ -48,23 +48,25 @@ before('Connecting to Selenium', function() {
 });
 
-beforeEach('Waiting for app to render', function() {
+beforeEach('Waiting for app to render', async function() {
```
There's no need to declare this as `async` because we're not `await`-ing anything, it's just returning a promise.

NiGhTTraX commented on tests/acceptance/setup.js:
```diff
@@ -48,23 +48,25 @@ before('Connecting to Selenium', function() {
-    coverage
+    coverage || ''
   );
+  if (!coverage) {
```
When would this happen? I would rather understand the problem than add warnings.

NiGhTTraX commented on tests/acceptance/specs/fixtures.spec.js:
```diff
@@ -1,23 +1,46 @@
+  return fixtureFiles.reduce(
+    (acc, f) => ({
+      ...acc,
+      [f.filePath]: importModule(require(f.filePath)) // eslint-disable-line
```
Please disable only the necessary rule.

NiGhTTraX commented on tests/acceptance/specs/fixtures.spec.js:
```diff
@@ -1,23 +1,46 @@
+      [f.filePath]: importModule(require(f.filePath)) // eslint-disable-line
+    }
+    ),
+    {}
```
These 3 lines can be joined :)

NiGhTTraX commented on tests/acceptance/specs/fixtures.spec.js:
```diff
@@ -1,23 +1,46 @@
+  );
+}
+
+describe('fixtures test', async() => {
```
I'm not sure if Mocha waits for this. As I recall defining tests needs to happen sync. Moreover, if it works, this is changing the previous behavior by introducing a prefix in the test names, thus forcing a rename of the screenshots.

Truth be told, I have the visual tests working with Cosmos@3 stashed somewhere and I just need to find the time to commit it. Since that's WIP and I'm fine with having failing acceptance tests in master I would rather not continue with making them work.

NiGhTTraX commented on tests/unit/components/docs-table/columns/default.spec.js:
```diff
@@ -8,4 +8,12 @@ describe('Default column', () => {
+    const $default = $render(<Default docs={{ required: true }} />);
+    expect($default.text()).to.equal('Required');
+  });
+  it('show default value if set', () => {
```
Please separate `it` blocks with a new line :)

NiGhTTraX commented on tests/unit/components/proxy.spec.js:
```diff
@@ -42,6 +46,15 @@ describe('DocsProxy', () => {
 
+    it('if no docs component passed, should render default', () => {
+      const FakeDocsTable = createSpy({ name: 'FakeDocsTable' });
+      createDocsProxy.__Rewire__('DocsTable', FakeDocsTable);
```
`DocsTable` being used by the proxy as a default value is an implementation detail - hence why something like rewire or enzyme is needed in the tests - so I don't want to unit test that. The acceptance tests will cover it by checking the whole thing end to end.

Commit:
fixing the code coverage for acceptance tests

NiGhTTraX reviewed (commented):
@scottdickerson thanks for all the hard work. Sorry to nitpick about the tests. In fact, I'm just gonna merge like this and worry about them another day. I need to bring in the latest changes from [react-test-buffet](https://github.com/NiGhTTraX/react-test-buffet) anyway.
